src/main/java/io/camunda/zeebe/bpmnassert/testengine/db/InMemoryZeebeDbTransactionContext.java

```diff
@@ -0,0 +1,46 @@
+package io.camunda.zeebe.bpmnassert.testengine.db;
+
+import io.camunda.zeebe.db.TransactionContext;
+import io.camunda.zeebe.db.TransactionOperation;
+import io.camunda.zeebe.db.ZeebeDbTransaction;
+import java.util.TreeMap;
+
+public class InMemoryZeebeDbTransactionContext implements TransactionContext {
+  private final InMemoryZeebeDbTransaction transaction;
+
+  public InMemoryZeebeDbTransactionContext(final TreeMap<Bytes, Bytes> database) {
+    transaction = new InMemoryZeebeDbTransaction(database);
+  }
+
+  @Override
+  public void runInTransaction(final TransactionOperation operations) {
+    try {
+      if (transaction.isInCurrentTransaction()) {
+        operations.run();
+      } else {
+        runInNewTransaction(operations);
+      }
+    } catch (final Exception e) {
+      throw new RuntimeException(
+          "Unexpected error occurred during zeebe db transaction operation.", e);
+    }
+  }
+
+  private void runInNewTransaction(final TransactionOperation operations) throws Exception {
+    try {
+      transaction.resetTransaction();
+      operations.run();
+      transaction.commit();
+    } finally {
+      transaction.rollback();
```

Comment:
❌ We should only rollback the transaction in case of an exception. If we've successfully committed there shouldn't be a need for this.

Reply:
Again, copied it from EZE. I agree with your point though. It would make more sense to call e.g. `transaction.reset();` (which internally does a rollback :-)

Reply:
How would you feel about just changing the `finally` into a `catch`? `commit()` already clears the caches after applying them to the database so it wouldn't be necessary to have a finally.